### PR TITLE
Feature: allowed experiment creators to set region enforced and enforced it in trials, added geolocation for experiments and trials, added geolocation warnings and provided a map of all geolocation experiment/trials

### PR DESCRIPTION
**Pull request by adamelamy**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features) **- manually, automated will be added for what is possible a bit later**
- [x] JavaDocs have been reviewed and added / updated if needed (for bug fixes / features) 
- [x] Code has been properly commented **- with JavaDocs**
- [x] Camel case has been used 

## Pull request type

 
**Note: Dependency changes are necessary**
 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Trial region box is useless and no enforcement existed. Experiment box took in a region, but it was a string and was also kind of useless.

Issue Number: All of 6

## What is the new behavior?

Users can create new experiments and check whether or not region should be enforced. If region is enforced then they must enter a region. This will also enforce it when adding a new trial within that particular experiment. Furthermore, when adding a region, the textbook is actually a button that goes to a map activity. In there, users can click to set down a marker or long click to delete their past markers. They can also just click elsewhere and their old marker will be deleted and moved to the new location. Finally, in the trial log when you click map, all the experiment's location (from the experiment and the trials) are shown. 

P.S. when getting region input the map starts at the user's current location if they allow it in permissions (there is a menu) - if they disallow it, we start at 0, 0. Additionally, when viewing all locations the map visibly pans to a particular locations. Additionally, note the toast messages that appear as they provide various pieces of information relevant to the specific context.

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

**Though any experiment that pre-existed my change needs to be deleted because their location is invalid.**

## Other information

I need @Jinlehs to find a way to re-include the back button in the Android menu bar, I was expecting to use that to allow the user to nicely cancel. I also just find it hard to navigate the app in general without it (especially the logs). Also, I would appreciate his expertise making my sh-*erhem* stuff look nicer, I find that on smaller screens everything in general tends to look a bit stuffy (including things I did not work  right now on like the logs). Finally, I find that the map stuff can be a bit slow, I have tried to pre-authenticate the app in main activity to somewhat mitigate that, but the more that can be improved the better.

### CODE REVIEW

File: code/app/src/main/java/com/cmput301w21t06/crowdfly/Views/Toaster.java [hunk 1]
```diff
@@ -3,8 +3,29 @@
...
+    }
+
+    /**
+     * This makes long lasting toasts i.e. extra crispy
+     * @param context
+     * This is the activity context
+     * @param s
+     * This is the string to depict
+     */
+    public static void makeCrispyToast(Context context, String s) {
```
Comment (qasimy123) [inline code/app/src/main/java/com/cmput301w21t06/crowdfly/Views/Toaster.java:28]: LOL

File: code/app/src/main/java/com/cmput301w21t06/crowdfly/Database/ExperimentController.java [hunk 2]
```diff
@@ -39,6 +39,7 @@
      * This sets up the snapshot listener for experiments
      */
     public static void setUp(){
+        Log.e("d","23اللعنة");
```
Comment (mharis2) [inline code/app/src/main/java/com/cmput301w21t06/crowdfly/Database/ExperimentController.java:42]: ممتاز

File: code/app/src/main/java/com/cmput301w21t06/crowdfly/Views/Toaster.java (hunk 1 above)
Comment (mharis2) [reply code/app/src/main/java/com/cmput301w21t06/crowdfly/Views/Toaster.java:28]: proper

**Review by mharis2 — APPROVED** [review]
ممتاز

File: code/app/src/main/java/com/cmput301w21t06/crowdfly/Database/ExperimentController.java (hunk 2 above)
Comment (adamelamy) [reply code/app/src/main/java/com/cmput301w21t06/crowdfly/Database/ExperimentController.java:42]: LOL

**Pull request merged**